### PR TITLE
Emit errors if TF or MHLO ops are found during StableHLO conversion.

### PR DESCRIPTION
**Pull request by ScottTodd**
This replaces https://github.com/openxla/iree/pull/16033. Progress on https://github.com/openxla/iree/issues/10297.

This would have given a better error message on https://github.com/openxla/iree/issues/16021 where the program had an unconverted TensorFlow op:
```mlir
...
    %85 = stablehlo.multiply %80, %83 : tensor<?x8x8x256xf32>
    %86 = stablehlo.compare  GT, %80, %84,  NOTYPE : (tensor<?x8x8x256xf32>, tensor<?x8x8x256xf32>) -> tensor<?x8x8x256xi1>
    %87 = stablehlo.select %86, %80, %85 : tensor<?x8x8x256xi1>, tensor<?x8x8x256xf32>
    %88 = "tf.ResizeNearestNeighbor"(%87, %3) {align_corners = false, device = "", half_pixel_centers = true} : (tensor<?x8x8x256xf32>, tensor<2xi32>) -> tensor<?x16x16x256xf32>
...
```
```
D:\dev\projects\iree (compiler-known-dialects)
(.venv) λ ..\iree-build\tools\iree-compile.exe --compile-to=input D:\dev\projects\iree-tmp\iree_unet_tf.mlir
D:\dev\projects\iree-tmp\iree_unet_tf.mlir:1:1: error: one or more illegal operations were found in the compiler input (are you missing an --iree-input-type= flag, or did you mean to pre-process through an IREE importer frontend?)
module {
^
D:\dev\projects\iree-tmp\iree_unet_tf.mlir:134:11: note: tf.ResizeNearestNeighbor
    %88 = "tf.ResizeNearestNeighbor"(%87, %3) {align_corners = false, device = "", half_pixel_centers = true} : (tensor<?x8x8x256xf32>, tensor<2xi32>) -> tensor<?x16x16x256xf32>
          ^
```

Note that this will _not_ give error messages if the _entire_ program is TF/MHLO - it must have _some_ StableHLO in it for the StableHLO input plugin code to run. In that example, the input type is "auto" (the default), which selects "stablehlo" based on the presence of stablehlo ops in the module.

### CODE REVIEW

File: compiler/plugins/input/StableHLO/stablehlo-iree/Conversion/VerifyCompilerInputLegality.cpp [hunk 1]
```diff
@@ -37,31 +37,51 @@ struct VerifyCompilerStableHloInputLegality final
...
     }
 
+    // Also check for dialects that should have been converted to stablehlo
+    // prior to being presented to IREE in the first place.
+    // We don't want to use source dependencies for these dialects, so we just
+    // use string matching.
+    llvm::DenseSet<StringRef> invalidDialects{"tf", "mhlo"};
+    llvm::DenseSet<Operation *> invalidOps;
+    getOperation().walk([&](Operation *op) {
+      if (invalidDialects.contains(op->getName().getDialectNamespace())) {
```
Comment (jpienaar) [inline compiler/plugins/input/StableHLO/stablehlo-iree/Conversion/VerifyCompilerInputLegality.cpp:60]: This is doing string comparisons, I'd rather check the IDs.

Comment (ScottTodd) [reply compiler/plugins/input/StableHLO/stablehlo-iree/Conversion/VerifyCompilerInputLegality.cpp:60]: Are the IDs available without a source dep on the TF / MHLO dialects?

See the comment above:
```
    // We don't want to use source dependencies for these dialects, so we just
    // use string matching.
```

**Comment by ScottTodd** [comment]
A check upstream in TensorFlow that it fully converts to stablehlo could also help here, if that does not already exist. torch-mlir has a similar check already (just via the conversion framework, no special pass needed).

Defense in depth :)

**Comment by benvanik** [comment]
+1 - we shouldn't have to check that the importer did it's job, as we should be a target for many things and we can't add checks for all of them (some we may not even know about).

**Comment by ScottTodd** [comment]
Okay, sounds like we should redirect issues with TF ops leaking through to the TF repo (or StableHLO). I'll close this PR.

**Pull request closed without merging**